onnxruntime/contrib_ops/cuda/custom_layer_norm.cu

```diff
@@ -0,0 +1,305 @@
+/* Modifications Copyright (c) Microsoft. */
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include <cooperative_groups.h>
+#include "custom_layer_norm.cuh"
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+namespace cg = cooperative_groups;
+using namespace onnxruntime::cuda;
+/*
+Custom kernel for Layer Normalization, adapted from https://github.com/microsoft/DeepSpeed.
+This layer requires template parameters since it relies fairly heavily on unrolling loops.
+The kernel is invoked when the n2(number of elements to reduce) is a multiple of a power of 2. 
+Default behavior for our purposes uses 256 threads for floats, and 128 threads for __half. 
+This restriction is a result of using the shift parameter to perform the minimum number of 
+register file shuffles necessary, which requires the number of threads in the secondary 
+reduction to be 1, 2, 4, 8, 16, or 32. The number of threads here corresponds to the number 
+of complete warps in the threadblock.
+For FP16, this kernel does not promote to FP32 in order to utilize the 2x throughput for
+__half2 instructions, and avoid the conversion overhead (1/8 of __hal2 arithmetic).
+For specific launch constraints, see the launch functions.
+*/
+
+template <int row_stride, int iterations>
+__global__ void custom_layer_norm(
+  float* vals,
+  const float* residual,
+  const float* gamma,
+  const float* beta,
+  float epsilon,
+  float* invvars = nullptr,
+  float* means = nullptr) {
+
+  constexpr int iteration_stride = row_stride / iterations;
+
+  cg::thread_block b = cg::this_thread_block();
+  cg::thread_block_tile<32> g = cg::tiled_partition<32>(b);
+
+  int row = blockIdx.x;
+  int id = threadIdx.x;
+  int gid = id / 32;
+
+  float vals_arr[iterations];
+  __shared__ float shr[iteration_stride >> 5];
+
+  float sum = 0.f;
+#pragma unroll
+  for (int i = 0; i < iterations; i++) {
+    vals_arr[i] = residual[row * row_stride + i * iteration_stride + id];
+    sum += vals_arr[i];
+  }
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride); 
+  }
+  
+  if (g.thread_rank() == 0) shr[gid] = sum;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> 5)) sum = shr[g.thread_rank()];
+
+#if __CUDA_ARCH__ < 700
+  b.sync();
+#endif
+
+  for (int stride = 1; stride < (iteration_stride >> 5); stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride); 
```

Comment:
just curious, why was it changed from g.shfl_down()?

Reply:
use of WARP_SHFL_DOWN allows us to switch between __shfl_down_sync(for compute capability 9.x) and __shfl_down(for lower versions), and is more consistent with the rest of the ORT codebase

Reply:
does the group's shfl_down() method also account for the different versions?
consistency is fine, but it's also good to use better tools/APIs if they are available. not saying this is better, i only just learned about cooperative_groups.h. it looks handy though.